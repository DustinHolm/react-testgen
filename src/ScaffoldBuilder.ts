import { createComponentRenderFunction } from "./templates/componentRenderingTemplate"
import { createDescribeBlock } from "./templates/describeBlockTemplate"
import { createDefaultImports, createImport } from "./templates/importTemplate"
import { createMock } from "./templates/mockTemplate"
import { createProp } from "./templates/propTemplate"
import { createResetMocksFunction } from "./templates/resetMocksTemplate"
import { Export, ImportBlock } from "./types"
import { getPropsForExport } from "./utils"

class ScaffoldBuilder {
    private name: string
    private exports: Export[]
    private imports: ImportBlock[]

    constructor(name: string) {
        this.name = name.replace(".tsx", "").replace(".ts", "")
        this.exports = []
        this.imports = []
    }

    addExport(element: Export) {
        this.exports.push(element)
    }

    addImport(element: ImportBlock) {
        this.imports.push(element)
    }

    build(): string {
        const jsxExports = this.exports.filter(e => e.element?.returnsJSX === true)
        const mainExport: Export | undefined = jsxExports.find(e => e.isDefault) ?? jsxExports[0]
        const props = mainExport !== undefined
            ? getPropsForExport(mainExport)
            : []

        const mocks = this.imports.filter(i => i.isInternal)

        const importsString = [
            createDefaultImports(),
            ...this.imports.map(i => createImport(i))
        ].join("\n")

        const describeString = createDescribeBlock(this.name)

        const resetMocksString = createResetMocksFunction(props, mocks)

        const mocksString = mocks
            .map(i => createMock(i))
            .join("\n")

        const propsString = props
            .map(p => createProp(p))
            .join("\n")

        const componentRenderString = mainExport !== undefined
            ? createComponentRenderFunction(mainExport.name, props)
            : ""

        return `\
${importsString}


${describeString}
${resetMocksString}
${mocksString}
${propsString}
${componentRenderString}
`
    }
}

export default ScaffoldBuilder