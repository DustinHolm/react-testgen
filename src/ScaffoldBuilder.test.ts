import ScaffoldBuilder from "./ScaffoldBuilder"

describe("ScaffoldBuilder Tests", () => {
    test("Matches snapshot", () => {
        const builder = new ScaffoldBuilder("Test.tsx")
        const returnedString = builder.build()
        expect(returnedString).toMatchSnapshot()
    })
})