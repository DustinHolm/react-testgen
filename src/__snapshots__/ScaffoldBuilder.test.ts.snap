// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScaffoldBuilder Tests Matches snapshot 1`] = `
"import { render, screen } from \\"@testing-library/react\\"


describe(\\"Test Tests\\", () => {
    beforeEach(() => {
        resetMocks()
    })

    test(\\"Component renders\\", () => {
        whenComponentIsRendered()
        screen.debug()
    })
})

function resetMocks() {

}




"
`;
